internal/certmonitor/main.go

```diff
@@ -0,0 +1,273 @@
+		if c.secretMatches(secret, monitor) {
+			c.findCertsFromSecret(secret)
+			break
+		}
+	}
+
+}
+
+func objectMatchesSelector(obj runtime.Object, selector *metav1.LabelSelector) bool {
+	if selector == nil {
+		return true
+	}
+	labelselector, err := metav1.LabelSelectorAsSelector(selector)
+	if err != nil {
+		return false
+	}
+
+	metaObj, ok := obj.(metav1.Object)
+	if !ok {
+		return false
+	}
+
+	return labelselector.Matches(labels.Set(metaObj.GetLabels()))
+
+}
+
+func (c *certMonitor) secretMatches(s *corev1.Secret, monitor MonitorConfig) bool {
+	if s == nil {
+		return false
+	}
+	if len(monitor.Secret.Name) > 0 && s.Name != monitor.Secret.Name {
+		return false
+	}
+	if len(monitor.Namespace.Name) > 0 && s.Namespace != monitor.Namespace.Name {
+		return false
+	}
+	if monitor.Secret.LabelSelector != nil && !objectMatchesSelector(s, monitor.Secret.LabelSelector) {
+		return false
+	}
+
+	if monitor.Namespace.LabelSelector != nil {
+		ns, err := c.namespaceLister.Get(s.Namespace)
+		if err != nil {
+			return false
+		}
+		if !objectMatchesSelector(ns, monitor.Secret.LabelSelector) {
+			return false
+		}
+	}
+	return true
+}
+
+func (c *certMonitor) findCertsFromSecret(secret *corev1.Secret) {
+	for dataKey, dataCert := range secret.Data {
+		certConv, err := base64.StdEncoding.DecodeString(string(dataCert))
+		if err != nil {
+			fmt.Printf("failed to decode base64 data: %v", err)
+			continue
+		}
+
+		pparse, _ := pem.Decode(certConv)
+		if pparse == nil {
+			fmt.Printf("failed to decode pem data on secret %s/%s/%s\n", secret.Namespace, secret.Name, dataKey)
+			continue
+		}
+
+		certss, err := x509.ParseCertificate(pparse.Bytes)
+		if err != nil {
+			fmt.Printf("failed to decode certificate on secret %s/%s/%s\n", secret.Namespace, secret.Name, dataKey)
+			continue
+		}
+
+		expiryTime := float64(certss.NotAfter.Unix())
+		certificatesExpiry.WithLabelValues(secret.Namespace, secret.Name, dataKey).Set(expiryTime)
+		fmt.Println("Certificates Expiry Date found:")
+		fmt.Println(certss.NotAfter)
+	}
+}
+
+//
+
+func (c *certMonitor) handleSecretCreation(obj interface{}) {
+	secret, ok := obj.(*corev1.Secret)
+	if !ok {
+		fmt.Printf("unexpected object: %v", obj)
+		return
+	}
+	fmt.Printf("Handling Creation: %s,%s\n", secret.Namespace, secret.Name)
+	c.findCertsFromSecret(secret)
+}
+
+func (c *certMonitor) handleSecretUpdate(oldObj, newObj interface{}) {
+	oldsecret, ok := oldObj.(*corev1.Secret)
+	if !ok {
+		fmt.Printf("unexpected object: %v", oldObj)
+		return
+	}
+
+	newsecret, ok := newObj.(*corev1.Secret)
+	if !ok {
+		fmt.Printf("unexpected object: %v", newObj)
+		return
+	}
+
+	if oldObj != nil {
+		for oldKey := range oldsecret.Data {
+			if _, ok := newsecret.Data[oldKey]; !ok {
+				certificatesExpiry.DeletePartialMatch(prometheus.Labels{
+					labelSecretName:      newsecret.Name,
+					labelSecretNamespace: newsecret.Namespace,
+					labelDataKey:         oldKey,
+				})
+			}
+		}
+	}
```

Comment:
Maybe you can exit early: 
```suggestion
Maybe you can exit early: 
```
        if oldSecret == nil || newSecret == nil {
            return
        }
        for oldKey := range oldsecret.Data {
	        if _, ok := newsecret.Data[oldKey]; !ok {
		        certificatesExpiry.DeletePartialMatch(prometheus.Labels{
			        labelSecretName:      newsecret.Name,
			        labelSecretNamespace: newsecret.Namespace,
			        labelDataKey:         oldKey,
		        })
	        }
        }
```